### DISCUSSION TITLE
Skoda Enyaq SoC

### DISCUSSION THREAD

**Discussion by stwa2020** [meatpiHQ/wican-fw#131] [Q&A]
Hi All,

I used Car Scanner to get the information for the SoC and have a question how to calculate the SoC based on the response,

Car Scanner: 82.29 % 
Tx : {"bus":"0","type":"tx","frame":[{"id":402391163,"dlc":8,"rtr":false,"extd":true,"data":[3,34,2,140,170,170,170,170]}]}
Rx: {"bus":"0","type":"rx","frame":[{"id":402522235,"dlc":8,"rtr":false,"extd":true,"data":[4,98,2,140,200,170,170,170]}]}
...
Car Scanner: 86.73 % 
Tx : {"bus":"0","type":"tx","frame":[{"id":402391163,"dlc":8,"rtr":false,"extd":true,"data":[3,34,2,140,170,170,170,170]}]}
Rx: {"bus":"0","type":"rx","frame":[{"id":402522235,"dlc":8,"rtr":false,"extd":true,"data":[4,98,2,140,210,170,170,170]}]}

The 5th value is the one for the battery and I created a list from 200 - 210 and the values which are reported in the Car Scanner.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 14">
<link id=Main-File rel=Main-File
href="file:///C:\Users\stefan\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\stefan\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>

Value 5 | Car Scanner % | delta
-- | -- | --
200 | 82,29 | 0,45
201 | 82,74 | 0,44
202 | 83,18 | 0,44
203 | 83,63 | 0,45
204 | 84,07 | 0,44
205 | 84,51 | 0,44
206 | 84,96 | 0,45
207 | 85,40 | 0,44
208 | 85,84 | 0,44
209 | 86,29 | 0,45
210 | 86,73 | 

</body>

</html>

Increasing the 5th value by one doesn't always add 0.44%, every third value adds 0.45%.

Question: Which Formula is used, is this a rounding done by Car Scanner and should I just use +0,444 for each increase for the 5th Value ?

Thanks and kind regards,

Stefan

**Comment by stwa2020** [comment]
I did some digging and created the following, based on the information I've found.

![image](https://github.com/meatpiHQ/wican-fw/assets/66866558/bd7c26a5-d321-4ec7-a7f0-e1a6121202b4)

![image](https://github.com/meatpiHQ/wican-fw/assets/66866558/44b1e5de-9e5c-42e0-ac29-732f860a2505)
![image](https://github.com/meatpiHQ/wican-fw/assets/66866558/93e828d5-6531-4ad2-beba-a45558f9beee)


can someone explain me how to continue with the type and conv to create the SoC of the Rx with 'Yellow' Value ?

**Reply by meatpiHQ** [reply]
So based on the SoC data you provided. The expression should be: (0.452*B4)-8.2

**Reply by stwa2020** [reply]
@meatpiHQ  
Thank you for your reply, I would like to understand your calculation, which number is B4 and how did you find the other numbers?
I will possibly discover and add other IDs and it would be nice to know were I can find this informations, an example would be quite useful.

Thanks, Stefan

**Reply by meatpiHQ** [reply]
@stwa2020 B4 represent the 4th byte(first byte is B0) that's you SoC. As for how I found the expression. It's a simple linear regression formula.

As for how to find other IDs, you can follow the example below. And what you did here is pretty good example of how to log the data and find the expression.

https://github.com/meatpiHQ/wican-fw/wiki/EV-Battery-SoC-in-Home-Assistant-%E2%80%90-Example

**Reply by stwa2020** [reply]
@meatpiHQ sorry to bother you again, when I use the number provided by CarScanner the -8.2 is changing and I would expect when the SoC is lower the difference will be higher, see the numbers for 76.09 % below.

Car Scanner show 76.09% and this is the Rx with B4=186
{"bus":"0","type":"rx","frame":[{"id":402522235,"dlc":8,"rtr":false,"extd":true,"data":[4,98,2,140,186,170,170,170]}]}

`(0,452*186)-8,20` = 75.872 %
The offset is not always -8.2 when B4=186 it's down to -7,98 
(`0.452*186)-7.98` =  76,092 % 

It looks like the second number (-8.2) is not linear, or is CarScanner doing the calculation in a 'different' way ? 

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 14">
<link id=Main-File rel=Main-File
href="file:///C:\Users\stefan\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\stefan\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>


B0 | B1 | B2 | B3 | B4 | B5 | B6 | B7 |   | CarScanner % |   | (0,452*B4) | -8,20 | Diff Car Scanner %
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
4 | 98 | 2 | 140 | 200 | 170 | 170 | 170 |   | 82,29 |   | 90,40 | 82,20 | 8,11
4 | 98 | 2 | 140 | 201 | 170 | 170 | 170 |   | 82,74 |   | 90,85 | 82,65 | 8,11
4 | 98 | 2 | 140 | 202 | 170 | 170 | 170 |   | 83,18 |   | 91,30 | 83,10 | 8,12
4 | 98 | 2 | 140 | 203 | 170 | 170 | 170 |   | 83,63 |   | 91,76 | 83,56 | 8,13
4 | 98 | 2 | 140 | 204 | 170 | 170 | 170 |   | 84,07 |   | 92,21 | 84,01 | 8,14
4 | 98 | 2 | 140 | 205 | 170 | 170 | 170 |   | 84,51 |   | 92,66 | 84,46 | 8,15
4 | 98 | 2 | 140 | 206 | 170 | 170 | 170 |   | 84,96 |   | 93,11 | 84,91 | 8,15
4 | 98 | 2 | 140 | 207 | 170 | 170 | 170 |   | 85,40 |   | 93,56 | 85,36 | 8,16
4 | 98 | 2 | 140 | 208 | 170 | 170 | 170 |   | 85,84 |   | 94,02 | 85,82 | 8,18
4 | 98 | 2 | 140 | 209 | 170 | 170 | 170 |   | 86,29 |   | 94,47 | 86,27 | 8,18
4 | 98 | 2 | 140 | 210 | 170 | 170 | 170 |   | 86,73 |   | 94,92 | 86,72 | 8,19
  |   |   |   |   |   |   |   |   |   |   |   |   |  
4 | 98 | 2 | 140 | 186 | 170 | 170 | 170 |   | 76,09 |   | 84,07 | 75,87 | 7,98



</body>

</html>
 

Thx, Stefan

**Comment by stwa2020** [comment]
@meatpiHQ I found it,

There are two values in CarScanner, the BMS % and the Display % but the ID and response is the same in CarScanner.
I took a closer look and found the following

**Display: 85.84%**
Tx : {"bus":"0","type":"tx","frame":[{"id":402391163,"dlc":8,"rtr":false,"extd":true,"data":[3,34,2,140,170,170,170,170]}]}
Rx: {"bus":"0","type":"rx","frame":[{"id":402522235,"dlc":8,"rtr":false,"extd":true,"data":[4,98,2,140,208,170,170,170]}]}

**BMS: 83.20%**
Tx: {"bus":"0","type":"tx","frame":[{"id":402391163,"dlc":8,"rtr":false,"extd":true,"data":[3,34,2,140,170,170,170,170]}]}"
Rx: {"bus":"0","type":"rx","frame":[{"id":402522235,"dlc":8,"rtr":false,"extd":true,"data":[4,98,2,140,208,170,170,170]}]}"

BMS = 208/2.5 = 83.20 %
Display = ??
I still don't know how the Display Value is calaculated ?

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 14">
<link id=Main-File rel=Main-File
href="file:///C:\Users\stefan\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\stefan\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>


  |   |   |   |   |   |   |   |   | Display | BMS |   | Calc Display | Calc BMS
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
B0 | B1 | B2 | B3 | B4 | B5 | B6 | B7 |   | CarScanner % | % |   | ?? | B4/2.5
4 | 98 | 2 | 140 | 200 | 170 | 170 | 170 |   | 82,29 |   |   |   | 80,00
4 | 98 | 2 | 140 | 201 | 170 | 170 | 170 |   | 82,74 |   |   |   | 80,40
4 | 98 | 2 | 140 | 202 | 170 | 170 | 170 |   | 83,18 |   |   |   | 80,80
4 | 98 | 2 | 140 | 203 | 170 | 170 | 170 |   | 83,63 | 81,20 |   |   | 81,20
4 | 98 | 2 | 140 | 204 | 170 | 170 | 170 |   | 84,07 | 81,60 |   |   | 81,60
4 | 98 | 2 | 140 | 205 | 170 | 170 | 170 |   | 84,51 | 82,00 |   |   | 82,00
4 | 98 | 2 | 140 | 206 | 170 | 170 | 170 |   | 84,96 | 82,40 |   |   | 82,40
4 | 98 | 2 | 140 | 207 | 170 | 170 | 170 |   | 85,40 | 82,80 |   |   | 82,80
4 | 98 | 2 | 140 | 208 | 170 | 170 | 170 |   | 85,84 | 83,20 |   |   | 83,20
4 | 98 | 2 | 140 | 209 | 170 | 170 | 170 |   | 86,29 |   |   |   | 83,60
4 | 98 | 2 | 140 | 210 | 170 | 170 | 170 |   | 86,73 |   |   |   | 84,00
  |   |   |   |   |   |   |   |   |   |   |   |   |  
4 | 98 | 2 | 140 | 186 | 170 | 170 | 170 |   | 76,09 |   |   |   | 74,40



</body>

</html>

**Comment by stwa2020** [comment]
I've done my tests, I suppose there is no easy formula to calculate the values displayed by CarScanner as SoC value, I've compared the CarScanner SoC value with the car display and it's not always correct, sometimes it rounds up and sometimes it rounds down.
However, there is a pattern that is repeated and can be used to get closer to the SoC value displayed in the car, below is the table which is based on CarScanner Values and MQTT received numbers, the display is not always switching when the B4 value changed it change after a few sec. that's the reason for *Nr* in the table.

Values may vary depending on battery size.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 14">
<link id=Main-File rel=Main-File
href="file:///C:\Users\stefan\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\stefan\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>


  |   |   |   |   |   |   |   |   | Display |   |   | Car Display |   | BMS |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
B0 | B1 | B2 | B3 | B4 | B5 | B6 | B7 |   | CarScanner % | diff |   | % |   | B4/2.5 | diff | Display - BMS
4 | 98 | 2 | 140 | 180 | 170 | 170 | 170 |   | 73,42 | 0,44 |   |   |   | 72,00 | 0,40 | 1,4200
4 | 98 | 2 | 140 | 181 | 170 | 170 | 170 |   | 73,86 | 0,45 |   |   |   | 72,40 | 0,40 | 1,4600
4 | 98 | 2 | 140 | 182 | 170 | 170 | 170 |   | 74,31 | 0,44 |   |   |   | 72,80 | 0,40 | 1,5100
4 | 98 | 2 | 140 | 183 | 170 | 170 | 170 |   | 74,75 | 0,44 |   |   |   | 73,20 | 0,40 | 1,5500
4 | 98 | 2 | 140 | 184 | 170 | 170 | 170 |   | 75,19 | 0,45 |   |   |   | 73,60 | 0,40 | 1,5900
4 | 98 | 2 | 140 | 185 | 170 | 170 | 170 |   | 75,64 | 0,44 |   |   |   | 74,00 | 0,40 | 1,6400
4 | 98 | 2 | 140 | 186 | 170 | 170 | 170 |   | 76,08 | 0,44 |   |   |   | 74,40 | 0,40 | 1,6800
4 | 98 | 2 | 140 | 187 | 170 | 170 | 170 |   | 76,52 | 0,45 |   |   |   | 74,80 | 0,40 | 1,7200
4 | 98 | 2 | 140 | 188 | 170 | 170 | 170 |   | 76,97 | 0,44 |   |   |   | 75,20 | 0,40 | 1,7700
4 | 98 | 2 | 140 | 189 | 170 | 170 | 170 |   | 77,41 | 0,44 |   |   |   | 75,60 | 0,40 | 1,8100
4 | 98 | 2 | 140 | 190 | 170 | 170 | 170 |   | 77,85 | 0,45 |   |   |   | 76,00 | 0,40 | 1,8500
4 | 98 | 2 | 140 | 191 | 170 | 170 | 170 |   | 78,30 | 0,44 |   |   |   | 76,40 | 0,40 | 1,9000
4 | 98 | 2 | 140 | 192 | 170 | 170 | 170 |   | 78,74 | 0,44 |   |   |   | 76,80 | 0,40 | 1,9400
4 | 98 | 2 | 140 | 193 | 170 | 170 | 170 |   | 79,18 | 0,45 |   |   |   | 77,20 | 0,40 | 1,9800
4 | 98 | 2 | 140 | 194 | 170 | 170 | 170 |   | 79,63 | 0,44 |   | 79 |   | 77,60 | 0,40 | 2,0300
4 | 98 | 2 | 140 | 195 | 170 | 170 | 170 |   | 80,07 | 0,44 |   | *80* |   | 78,00 | 0,40 | 2,0700
4 | 98 | 2 | 140 | 196 | 170 | 170 | 170 |   | 80,51 | 0,45 |   | 80 |   | 78,40 | 0,40 | 2,1100
4 | 98 | 2 | 140 | 197 | 170 | 170 | 170 |   | 80,96 | 0,44 |   | *81* |   | 78,80 | 0,40 | 2,1600
4 | 98 | 2 | 140 | 198 | 170 | 170 | 170 |   | 81,40 | 0,44 |   | 81 |   | 79,20 | 0,40 | 2,2000
4 | 98 | 2 | 140 | 199 | 170 | 170 | 170 |   | 81,84 | 0,45 |   | *82* |   | 79,60 | 0,40 | 2,2400
4 | 98 | 2 | 140 | 200 | 170 | 170 | 170 |   | 82,29 | 0,44 |   | 82 |   | 80,00 | 0,40 | 2,2900
4 | 98 | 2 | 140 | 201 | 170 | 170 | 170 |   | 82,74 | 0,44 |   | *83* |   | 80,40 | 0,40 | 2,3400
4 | 98 | 2 | 140 | 202 | 170 | 170 | 170 |   | 83,18 | 0,45 |   | 83 |   | 80,80 | 0,40 | 2,3800
4 | 98 | 2 | 140 | 203 | 170 | 170 | 170 |   | 83,63 | 0,44 |   | 83 |   | 81,20 | 0,40 | 2,4300
4 | 98 | 2 | 140 | 204 | 170 | 170 | 170 |   | 84,07 | 0,44 |   | *84* |   | 81,60 | 0,40 | 2,4700
4 | 98 | 2 | 140 | 205 | 170 | 170 | 170 |   | 84,51 | 0,45 |   | 84 |   | 82,00 | 0,40 | 2,5100
4 | 98 | 2 | 140 | 206 | 170 | 170 | 170 |   | 84,96 | 0,44 |   | *85* |   | 82,40 | 0,40 | 2,5600
4 | 98 | 2 | 140 | 207 | 170 | 170 | 170 |   | 85,40 | 0,44 |   | 85 |   | 82,80 | 0,40 | 2,6000
4 | 98 | 2 | 140 | 208 | 170 | 170 | 170 |   | 85,84 | 0,45 |   | *86* |   | 83,20 | 0,40 | 2,6400
4 | 98 | 2 | 140 | 209 | 170 | 170 | 170 |   | 86,29 | 0,44 |   | 86 |   | 83,60 | 0,40 | 2,6900
4 | 98 | 2 | 140 | 210 | 170 | 170 | 170 |   | 86,73 | 0,44 |   | *87* |   | 84,00 | 0,40 | 2,7300
4 | 98 | 2 | 140 | 211 | 170 | 170 | 170 |   | 87,17 | 0,45 |   |   |   | 84,40 | 0,40 | 2,7700
4 | 98 | 2 | 140 | 212 | 170 | 170 | 170 |   | 87,62 | 0,44 |   |   |   | 84,80 | 0,40 | 2,8200
4 | 98 | 2 | 140 | 213 | 170 | 170 | 170 |   | 88,06 | 0,44 |   |   |   | 85,20 | 0,40 | 2,8600
4 | 98 | 2 | 140 | 214 | 170 | 170 | 170 |   | 88,50 | 0,45 |   |   |   | 85,60 | 0,40 | 2,9000
4 | 98 | 2 | 140 | 215 | 170 | 170 | 170 |   | 88,95 | 0,44 |   |   |   | 86,00 | 0,40 | 2,9500
4 | 98 | 2 | 140 | 216 | 170 | 170 | 170 |   | 89,40 | 0,44 |   |   |   | 86,40 | 0,40 | 3,0000
4 | 98 | 2 | 140 | 217 | 170 | 170 | 170 |   | 89,84 | 0,45 |   |   |   | 86,80 | 0,40 | 3,0400
4 | 98 | 2 | 140 | 218 | 170 | 170 | 170 |   | 90,29 | 0,44 |   |   |   | 87,20 | 0,40 | 3,0900
4 | 98 | 2 | 140 | 219 | 170 | 170 | 170 |   | 90,73 | 0,44 |   |   |   | 87,60 | 0,40 | 3,1300
4 | 98 | 2 | 140 | 220 | 170 | 170 | 170 |   | 91,17 | 0,45 |   |   |   | 88,00 | 0,40 | 3,1700



</body>

</html>

**Reply by stwa2020** [reply]
According to this calculation,
CarScanner will show 0.27% SoC if B4 is 15 and BMS shows 6%.
CarScanner will show 100.05% SoC if B4 is 242 and BMS shows 96%.
Unknow which % will be shown on the Car Display and if B4 can be higher than 242.

**Reply by stwa2020** [reply]
ODB2 will create the BMS SoC and not the SoC used in the App or Car Display, B4/2.5 will not create the App or Car Display SoC. Car Scanner shows almost the correct values, it is more or less an approximation.
Comparing CarScanner and ODB2 value B4 produces the following diagram.

![image](https://github.com/meatpiHQ/wican-fw/assets/66866558/0b8ca41f-40c9-4815-a1d0-5f15d59ec55a)

**Reply by meatpiHQ** [reply]
I think the best way is to use the car display the source of truth, if you use this expression it will give you an approximate value.

Expression: (0.46*B4)−10.08

**Reply by stwa2020** [reply]
Thank you for the expression, I have created a script in IOBroker bockly based on the CarScanner calculation which is 99.5% the number that shows the car display within a 5min interval when it is above 50%, I need to check if the value <50% is also using the same calculation.

I've added the expression to the diagram, it's easy to get the BMS Value but it's another Story, for the SoC which is shown on the Display and in the App.

![image](https://github.com/meatpiHQ/wican-fw/assets/66866558/4db35479-dbf2-461c-a950-1a019530867a)
